Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder.py

```diff
@@ -0,0 +1,614 @@
+
+# Files ending in -ref/-notref aren't considered expectations, but are skipped.
+skipped_test_suffixes = tuple(
+    "".join(pair)
+    for pair in itertools.product(
+        ("-expected", "-expected-mismatch", "-ref", "-notref"),
+        supported_reference_extensions,
+    )
+)
+
+digit_re = re.compile("([0-9]+)")
+
+
+def natsort(string_to_split):
+    split = digit_re.split(string_to_split)
+    # this converts the split numbers into tuples so that "01" < "1"
+    split[1::2] = [(int(i), i) for i in split[1::2]]
+    return split
+
+
+class LayoutTestFinder(object):
+    def __init__(self, fs, layout_tests_base_dir, baseline_search_paths):
+        """Find layout tests.
+
+        :param FileSystem fs: the current filesystem object
+        :param str layout_tests_base_dir: the base directory of LayoutTests
+            (see: Port.layout_tests_dir())
+        :param List[str] baseline_search_paths: the baseline search paths, from most to
+            least specific (see: Port.baseline_search_path(device_type))
+        """
+        self.fs = fs
+
+        layout_tests_base_dir = fs.normpath(fs.realpath(layout_tests_base_dir))
+        baseline_search_paths = [
+            fs.normpath(fs.realpath(bsp)) for bsp in baseline_search_paths
+        ]
+
+        self.layout_tests_base_dir = layout_tests_base_dir
+        self.baseline_search_paths = baseline_search_paths
+
+        # XXX: both this and the realpath calls above mean __init__ reads the
+        # filesystem, and object data reflects FS state this time; are we okay with
+        # this?
+        self.w3c_support_dirs, self.w3c_support_files = self._load_w3c_resource_data()
+
+    def _load_w3c_resource_data(self):
+        w3c_path = self.fs.join(
+            self.layout_tests_base_dir,
+            "imported",
+            "w3c",
+        )
+
+        if not self.fs.exists(w3c_path):
+            return {}, {}
+
+        path = self.fs.join(
+            w3c_path,
+            "resources",
+            "resource-files.json",
+        )
+
+        with self.fs.open_binary_file_for_reading(path) as fd:
+            json_data = json.load(fd)
+
+        dirs_by_parent = {}
+        for fullname in json_data["directories"]:
+            dirname, basename = fullname.rsplit("/", 1)
+            dirs_by_parent.setdefault(
+                ("imported/w3c/" + dirname).replace("/", self.fs.sep), set()
+            ).add(basename)
+
+        files_by_parent = {}
+        for fullname in json_data["files"]:
+            dirname, basename = fullname.rsplit("/", 1)
+            files_by_parent.setdefault(
+                ("imported/w3c/" + dirname).replace("/", self.fs.sep), set()
+            ).add(basename)
+
+        return dirs_by_parent, files_by_parent
+
+    def _canonicalize_test_path(self, path, base):
+        fs = self.fs
+
+        base = fs.realpath(base)
+        path = fs.realpath(path)
+
+        relpath = fs.relpath(path, base)
+        if relpath.startswith(".." + fs.sep):
```

Comment:
Is this the only case we want to convert to an `abspath`?

Reply:
I think so? Can you think of any other?